frontend/catalyst/jax_primitives.py

```diff
@@ -1634,12 +1624,10 @@ def _var_lowering(jax_ctx: mlir.LoweringRuleContext, obs: ir.Value, shots: int,
 def _probs_abstract_eval(obs, shape, shots=None):
     assert isinstance(obs, AbstractObs)
 
-    if obs.primitive is compbasis_p:
-        assert shape == (2**obs.num_qubits,)
-    else:
+    if obs.primitive is not compbasis_p:
         raise TypeError("probs only supports computational basis")
 
-    return core.ShapedArray(shape, jax.numpy.float64)
+    return core.DShapedArray(shape, np.dtype("float64"))
```

Comment:
Note that I don't think it will be sufficient to just change the `ShapedArray` to `DShapedArray`, we want to make sure that the relevant machinery will be okay with a measurement process producing a dynamically-shaped tensor instead of a statically-shaped one. 
So we'll want to trace back where the `shape` comes from, does that mechanism work if at the source the number of qubits was now a tracer? Do the observables work with it?
Further down, does the MLIR generation work (i.e. the lowering aspect of these primitives), do the operations of the quantum dialect work with it, etc.

At the moment we will always generate a static value for the `shape` parameter here, so our current tests will not properly test this. We'll want to write some artificial (unit) tests where we make sure each relevant function will work with a dynamic shape for the measurement processes.

Reply:
I agree. This PR just tackle the frontend part. I will address the MLIR part in a subsequent PR.

Also, I have added the artificial tests. Done :)